src/rust/engine/process_execution/sandboxer/src/lib.rs

```diff
@@ -297,3 +321,173 @@ impl SandboxerClient {
+        if let Some(proc) = locked_proc.as_mut() {
+            proc.process.check_child_has_exited().map(|b| !b)
+        } else {
+            Ok(false)
+        }
+    }
+
+    pub async fn spawn(&self) -> Result<(), String> {
+        let mut locked_proc = self.process.write().await;
+        self.do_spawn(&mut locked_proc).await
+    }
+
+    async fn do_spawn(
+        &self,
+        locked_proc: &mut WriteLockedSandboxerProcess<'_>,
+    ) -> Result<(), String> {
+        ensure_parent_dir(&self.log_path)?;
+        let logfile = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(&self.log_path)
+            .await
+            .map_err(|e| e.to_string())?;
+        let mut cmd = Command::new(&self.sandboxer_bin);
+        cmd.arg("--socket-path");
+        cmd.arg(self.socket_path.as_os_str());
+        cmd.args(self.store_cli_opt.to_cli_args());
+        // Pass the calling code's max log level to the sandboxer process.
+        cmd.env("RUST_LOG", log::max_level().as_str());
+        cmd.stderr(Stdio::from(logfile.into_std().await));
+
+        debug!(
+            "Spawning sandboxer with cmd: {} {}",
+            &self.sandboxer_bin.to_string_lossy(),
+            cmd.as_std()
+                .get_args()
+                .map(|a| a.to_string_lossy())
+                .collect::<Vec<_>>()
+                .join(" ")
+        );
+
+        if locked_proc.is_none() {
+            let process = ManagedChild::spawn(&mut cmd, Some(Duration::from_millis(1000)))
+                .map_err(|e| e.to_string())?;
+            debug!("Spawned sandboxer process {:?}", process.id());
+            let client = SandboxerClient::connect_with_retries(&self.socket_path).await?;
+            let _ = locked_proc.insert(SandboxerProcess { process, client });
+        }
+        Ok(())
+    }
+
+    async fn kill(&self) {
+        let mut locked_proc = self.process.write().await;
+        self.do_kill(&mut locked_proc).await
+    }
+
+    async fn do_kill(&self, locked_proc: &mut RwLockWriteGuard<'_, Option<SandboxerProcess>>) {
+        // Note that we don't want to remove the socket file here, as it may no longer
+        // be under the control of the sandboxer process we're killing.
+        if let Some(ref mut proc) = **locked_proc {
+            // Best effort to shut down, ignoring any errors (which would
+            // typically be due to the process already being dead).
+            let _ = proc.process.attempt_shutdown_sync();
+        }
+        locked_proc.take();
+    }
+
+    async fn client(&self) -> Result<SandboxerClient, String> {
+        {
+            let proc = self.process.read().await;
+            if let Some(proc) = proc.as_ref() {
+                return Ok(proc.client.clone());
+            }
+        }
+        // We've released the read lock, so now we can acquire the write lock.
+        let mut locked_proc = self.process.write().await;
```

Comment:
Is the read/write lock necessary? Why not just use a regular mutex and spawn the process under the mutex (without dropping a read lock and reacquiring a write lock) if the process is not running?

Reply:
Because we want multiple concurrent "readers" to be able to send RPCs to a running sandboxer. We only want to block those "readers" and take exclusive control when we have to respawn.

Reply:
But the read lock is not kept for the entire RPC call, just to clone the client. So multiple clients exist without the read lock being held. Presumably, a mutex would be held for a short period of time to clone the client as well and then unlocked.

Reply:
Hmm, yes, good point. Let me chew on that.

Reply:
Going the other way - should we hold a shared lock while executing the RPC call? That would prevent the server being killed while requests were outstanding. But I'm not sure that's important - the server is unlikely to be killed in this way, and we have to be robust to a server being killed by some external process beyond our control anyway.

Reply:
> I think this would work with a Mutex, at the (probably low) risk of unnecessary serialization if that mutex were highly contended. But what is the advantage? I guess the code would be slightly simpler.

Mainly it would be easier to understand. If the lock was highly contended and the process needed to be restarted, currently a bunch of readers will drop the read lock and contend on the write lock even after the first writer to win restarts the process. 

With a Mutex, you would still have contention on the lock but no complexity of switching from read lock to write lock.

Reply:
> Going the other way - should we hold a shared lock while executing the RPC call? That would prevent the server being killed while requests were outstanding. But I'm not sure that's important - the server is unlikely to be killed in this way, and we have to be robust to a server being killed by some external process beyond our control anyway.

Probably not. We never did that in any toolchain code for gRPC servers. I would check whether Hyper contains graceful shutdown code which allows existing in-flight requests to finish.

Reply:
Well, at toolchain we never had gRPC clients that could restart their corresponding servers, effectively owning them. This is a very unique circumstance.

The existing graceful shutdown code does work (it sends an HTTP/2 GOAWAY frame, which the client respects). But this only works if the unix domain socket is still viable.

Reply:
> > I think this would work with a Mutex, at the (probably low) risk of unnecessary serialization if that mutex were highly contended. But what is the advantage? I guess the code would be slightly simpler.
> 
> Mainly it would be easier to understand. If the lock was highly contended and the process needed to be restarted, currently a bunch of readers will drop the read lock and contend on the write lock even after the first writer to win restarts the process.

Sure, but the case I'm thinking of is the usual one, where the process doesn't need restarting. The current code will have no contention, but a Mutex could.

Reply:
I'm still of two minds about the Mutex. It would make the code a bit simpler, and is pretty unlikely to create a problem in practice... I *think*. OTOH, why introduce even the potential for contention... I'll chew on this for a bit and either change or merge. Thanks!